### DISCUSSION TITLE
A2DP Stream Audio to Bluetooth Speaker ... Control

### DISCUSSION THREAD

**Discussion by HobmanX** [pschatzmann/arduino-audio-tools#1506] [Q&A] [answered]
I am sending audio to a JBL GO blutooth speaker that has volume control buttons on it.
The buttonson the speaker work for analog audio in via the s.5mm audio jack and work for Bluetooth audio from my phone to the speaker.
When I run my sketch in ESP32-PICO-D4..... ( a few lines of the bluetooth bit sre shown below)..... The audio is sent ok but it is at what ever volume it is with the control in the sketch. Pressing the buttons (long or short) makes no difference. I can alter the volume by changing the volume setting/config in the sketch code but that means more pins/buttons on my project board and it is not controlled locally by the listener at the speaker
```

const char* name = "JBL GO";       // Replace with your bluetooth speaker name  

BluetoothA2DPSource a2dp_source;      // A2DP Sender

// callback used by A2DP to provide the sound data - usually len is 128 * 2 channel int16 frames
int32_t get_sound_data(uint8_t * data, int32_t len) {
  return resample1.readBytes((uint8_t*)data, len);
}


//this bit below is in setup....

  // start the bluetooth
  Serial.println("starting A2DP...");
  a2dp_source.set_auto_reconnect(false);
  a2dp_source.start_raw(name, get_sound_data);
 
//this bit above is in setup....

```

**Comment by pschatzmann** [comment]
Did you try to set the log level to debug and check
- if any events are received when you change the volume
- why the volume is not updated ?

**Reply by HobmanX** [reply]
I think I may have a memory size problem in ESP32-PICO-D4
When I uncomment the following 2 line in sketch   ........... 
//  Serial.begin(115200);
//  AudioLogger::instance().begin(Serial, AudioLogger::Debug);
I only get short bursts of audio and don't see any activity on the terminal when buttons presses
Sketch uses 1300329 bytes (99%) of program storage space. Maximum is 1310720 bytes.
Global variables use 47168 bytes (14%) of dynamic memory, leaving 280512 bytes for local variables. Maximum is 327680 bytes.
//xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
Wth the lines commented out I get the following when uploading and get BT connected and audio continuous tone.
Sketch uses 1282525 bytes (97%) of program storage space. Maximum is 1310720 bytes.
Global variables use 47040 bytes (14%) of dynamic memory, leaving 280640 bytes for local variables. Maximum is 327680 bytes.

//xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

This is the messages from the Debug  screens......... don't know what they mean..????

Waiting for gdb server to start...[2024-04-14T09:46:39.122Z] SERVER CONSOLE DEBUG: onBackendConnect: gdb-server session connected. You can switch to "DEBUG CONSOLE" to see GDB interactions.
"C:\\Users\\abc\\AppData\\Local\\Arduino15\\packages\\esp32\\tools\\openocd-esp32\\v0.11.0-esp32-20221026/bin/openocd" -c "gdb_port 50000" -c "tcl_port 50001" -c "telnet_port 50002" -s "d:\\abc\\Documents\\Arduino\\Sketchbook\\ESPipXResVolMixMar24" -f "C:/Program Files/Arduino IDE/resources/app/plugins/cortex-debug/extension/support/openocd-helpers.tcl" -f debug.cfg
Open On-Chip Debugger v0.11.0-esp32-20221026 (2022-10-26-14:48)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
CDRTOSConfigure
Info : Listening on port 50001 for tcl connections
Info : Listening on port 50002 for telnet connections
Error: no device found
Error: unable to open ftdi device with vid 0403, pid 6010, description '*', serial '*' at bus location '*'
Error: no device found
Error: unable to open ftdi device with vid 0403, pid 6014, description '*', serial '*' at bus location '*'

[2024-04-14T09:46:39.291Z] SERVER CONSOLE DEBUG: onBackendConnect: gdb-server session closed
GDB server session ended. This terminal will be reused, waiting for next session to start...

//xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
So is the A2DP stream protocol supposed to receive control button messages from speaker and change the volume values in the Sketch........?   Or does this have to be programmed into the sketch to be ready for it....???

PS...... I tried another Bluetooth Speaker and i get audio out and it does control volume from the buttons on the speaker.
So some speakers must have local control and some don't............but I dont get any activity on Arduino screens so must be controlling just within the speaker....?

**Reply by pschatzmann** [reply, marked as answer]
If log level debug does not help try it with info.
You want to see which callback methods that are called when you change the volume (if any), so that you know what code to double check.

Yes, there is some logic in place, but since I do not have such a speaker, it has never been tested and maybe some steps are missing.
